trie/encoding.go

```diff
@@ -26,43 +28,82 @@ package trie
 // contains a value. Hex key encoding is used for nodes loaded in memory because it's
 // convenient to access.
 //
-// COMPACT encoding is defined by the Ethereum Yellow Paper (it's called "hex prefix
-// encoding" there) and contains the bytes of the key and a flag. The high nibble of the
-// first byte contains the flag; the lowest bit encoding the oddness of the length and
-// the second-lowest encoding whether the node at the key is a value node. The low nibble
-// of the first byte is zero in the case of an even number of nibbles and the first nibble
-// in the case of an odd number. All remaining nibbles (now an even number) fit properly
-// into the remaining bytes. Compact encoding is used for nodes stored on disk.
-
-func hexToCompact(hex []byte) []byte {
-	terminator := byte(0)
-	if hasTerm(hex) {
-		terminator = 1
-		hex = hex[:len(hex)-1]
+// BINARY COMPACT ENCODING FORMAT:
+// Header Nibble:
+// - first bit: 1 if should be terminated / 0 if not,
+// - bits 2-4: the number of unused (least significant) bits from last byte = [8 - ((4 (header nibble) + message len) % 8)] % 8
+// Body:
+// -- bits 5-8 of first byte and all subsequent bytes: tightly-packed key bits without terminator
+// -- bits 2-8 of last byte (varies based on key length) 0 bits of padding
+//
+// Example:
+// 1 1 0 1 1 2(terminator)
```

Comment:
Yes, this is an example binary key, where each bit is represented by a full byte. This is what makes 2 a valid terminator. Since I'm going to basically completely re-write this file, I'll make the docs better as well 👍